### DISCUSSION TITLE
Can we read/write to a folder? How?

### DISCUSSION THREAD

**Discussion by tleylan** [vercel/community#241] [Help] [answered]
I'm trying to read/write to a known folder in my app. It works on my local machine because I created the folder. How do I create such a folder on a deployed app? Thanks.

To clarify. I can check for the existence of a folder and create it if it doesn't I suppose I'm asking if that is the only method? Is there a way to initialize such a folder outside of the app as well?

**Comment by styfle** [comment, marked as answer]
Its generally considered an anti-pattern to write files from your Serverless Functions because you can't guarantee the next invocation is the same function so the file might be missing.

Its best to keep your functions stateless and store shared state in [redis or a database](https://vercel.com/docs/solutions/databases).

If you have a dependency that must write files, you can use `require('os').tmpdir()` which should be `/tmp`.

**Reply by tleylan** [reply]
I did get it working but ultimately did not like the solution. I've done it with SQLite3 now and I'm much happier. Thanks much.

**Reply by hirehamir** [reply]
`require('os').tmpdir()`, or the `import { tmpdir } from 'os';` EcmaScript (ES) module equivalent, helped me get rid of the `EROFS: read-only file system, mkdir` error I was receiving from Vercel.

It might be worth emphasizing this more in the documentation, since I had to go through a couple of search engine links to find this workaround.

Thank you for posting this fix!

**Comment by tleylan** [comment]
Bummer... I spoke a bit too soon. Everything works locally but I'm getting a database not found error when deployed. I can't seem to find a way to view the folders but it should have copied the database file. Possibly I need to define that step somewhere in the deployment. Any pointers on this? Thank you.

**Reply by styfle** [reply]
Are you reading the file like this?

```js
const { readFileSync } = require('fs');
const { join } = require('path');
const file = readFileSync(join(__dirname, 'folder', 'file.txt'), 'utf8');

module.exports = (req, res) => {
  res.end(file);
}
```

**Reply by tleylan** [reply]
Hi again. No I stopped trying to use files that wasn't ultimately going to work. I switched to a database using SQLite3. This runs on my local machine but gives me an error when I try to use it deployed on Vercel. The error is related to not finding the database file and so it can't connect. The database file should be present (I think) but perhaps it isn't. I don't seem to have any way to check the folder structure post deployment (or do I)?

**Reply by styfle** [reply]
It’s hard to say what the problem is without any code. 

Your SQLite DB is going to be read only so it’s probably not the best solution for serverless (see my first post). 

It you want a read only local data tatie, you could use a json file instead of a binary SQLite file and read it like any other file. 

You can “check” the file system post deploy by using fs.readdir() but there is no UI to do this.

**Reply by tleylan** [reply]
Ok, thanks again. I believe I've misunderstood a bit about what one can do with Vercel and/or how to actually do it. I was hoping to have it host a simple API accessible from a VR metaverse (the VR part works). It could work with a flat file of some sort but clearly  the addition of SQL would make it far more flexible. Believe I will have to formulate a Plan B :-)

**Reply by mmahalwy** [reply]
@tleylan im running into the same exact problem. Did you ever find a solution to this?

**Reply by tleylan** [reply]
I implemented Plan B and went to another server that I host my rest-based services on.

**Reply by hirehamir** [reply]
Out of curiosity, what are you using to host that server, and, if you don't mind me asking, how much does it cost?

I've been looking for zero-cost servers for prototyping but they typically have a lot of restrictions.

**Reply by tleylan** [reply]
I have a virtual server from HyperFive https://www.hyperfive.com. It's nice to have the freedom to install/setup what you want. The cost will vary depending on the system you need but you can see for experimentation the costs can be quite minimal.
